### PR TITLE
statistics: correct behavior of non-lite InitStats and stats sync load of no stats column

### LINKED ISSUE #57804: Too many unnecessary log for sync load when stats is not exists
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

many sysbench tables and run read_write
tidb always print the log but the column `pad` stats is not collected.
```
[2024/11/28 11:56:52.620 +00:00] [WARN] [stats_syncload.go:412] ["fail to get hist meta for this histogram, possibly a deleted one"] [table_id=1285152] [hist_id=3] [is_index=false]
[2024/11/28 11:56:52.623 +00:00] [WARN] [stats_syncload.go:412] ["fail to get hist meta for this histogram, possibly a deleted one"] [table_id=1201116] [hist_id=3] [is_index=false]
[2024/11/28 11:56:52.624 +00:00] [WARN] [stats_syncload.go:412] ["fail to get hist meta for this histogram, possibly a deleted one"] [table_id=1306426] [hist_id=3] [is_index=false]
[2024/11/28 11:56:52.629 +00:00] [WARN] [stats_syncload.go:412] ["fail to get hist meta for this histogram, possibly a deleted one"] [table_id=1357610] [hist_id=3] [is_index=false]
[2024/11/28 11:56:52.634 +00:00] [WARN] [stats_syncload.go:412] ["fail to get hist meta for this histogram, possibly a deleted one"] [table_id=1236468] [hist_id=3] [is_index=false]
[2024/11/28 11:56:52.634 +00:00] [WARN] [stats_syncload.go:412] ["fail to get hist meta for this histogram, possibly a deleted one"] [table_id=1393198] [hist_id=3] [is_index=false]
...
```
### 4. What is your TiDB version? (Required)

Release Version: v8.5.0-alpha-232-g924784aa3e
Edition: Community
Git Commit Hash: 924784aa3e57db874633eff00f435586eaf05bd8
Git Branch: HEAD
UTC Build Time: 2024-11-28 02:37:38
GoVersion: go1.23.3
Race Enabled: false

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close #57804

Problem Summary:

### What changed and how does it work?

You can regard this pull request as a totally rewrite of https://github.com/pingcap/tidb/pull/53399 and https://github.com/pingcap/tidb/pull/53298

This pull reduced the possible stats state after the stats initialization finished. Now we'll get the same memory objects after a non-concurrent or current non-lite load.

Then use the unified status to make the `ColumnIsLoadNeeded` correct when one column even has no record in mysql.stats_histograms.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/statistics/handle/handletest/statstest/stats_test.go [hunk 1]
```diff
@@ -366,16 +366,10 @@ func initStatsVer2(t *testing.T, isConcurrency bool) {
...
-		require.Equal(t, uint8(0x33), table0.GetCol(1).LastAnalyzePos.GetBytes()[0])
-		require.Equal(t, uint8(0x33), table0.GetCol(2).LastAnalyzePos.GetBytes()[0])
-		require.Equal(t, uint8(0x33), table0.GetCol(3).LastAnalyzePos.GetBytes()[0])
-		require.Equal(t, uint8(0x3), table0.GetIdx(1).LastAnalyzePos.GetBytes()[0])
-		require.Equal(t, uint8(0x3), table0.GetIdx(2).LastAnalyzePos.GetBytes()[0])
-	}
+	require.Equal(t, 0, table0.ColNum())
+	require.Equal(t, 2, table0.IdxNum())
+	require.Equal(t, uint8(0x3), table0.GetIdx(1).LastAnalyzePos.GetBytes()[0])
+	require.Equal(t, uint8(0x3), table0.GetIdx(2).LastAnalyzePos.GetBytes()[0])
```
Comment (winoros) [inline pkg/statistics/handle/handletest/statstest/stats_test.go:372]: Both single-thread and concurrently init now get the same result. More consistent than before.

File: pkg/statistics/handle/handletest/statstest/stats_test.go [hunk 2]
```diff
@@ -366,16 +372,15 @@ func initStatsVer2(t *testing.T, isConcurrency bool) {
...
-	if isConcurrency {
-		require.Equal(t, uint8(0x3), table0.GetIdx(1).LastAnalyzePos.GetBytes()[0])
-		require.Equal(t, uint8(0x3), table0.GetIdx(2).LastAnalyzePos.GetBytes()[0])
-	} else {
-		require.Equal(t, uint8(0x33), table0.GetCol(1).LastAnalyzePos.GetBytes()[0])
-		require.Equal(t, uint8(0x33), table0.GetCol(2).LastAnalyzePos.GetBytes()[0])
-		require.Equal(t, uint8(0x33), table0.GetCol(3).LastAnalyzePos.GetBytes()[0])
-		require.Equal(t, uint8(0x3), table0.GetIdx(1).LastAnalyzePos.GetBytes()[0])
-		require.Equal(t, uint8(0x3), table0.GetIdx(2).LastAnalyzePos.GetBytes()[0])
-	}
```
Comment (winoros) [inline pkg/statistics/handle/handletest/statstest/stats_test.go:378]: In fact, we should not differ the single thread init from the concurrent init.
This pr makes them consistent.

File: pkg/statistics/handle/handletest/statstest/stats_test.go [hunk 3]
```diff
@@ -366,16 +372,15 @@ func initStatsVer2(t *testing.T, isConcurrency bool) {
...
+	require.Equal(t, 5, table0.ColNum())
+	require.True(t, table0.GetCol(1).IsAllEvicted())
+	require.True(t, table0.GetCol(2).IsAllEvicted())
+	require.True(t, table0.GetCol(3).IsAllEvicted())
+	require.True(t, !table0.GetCol(4).IsStatsInitialized())
+	require.True(t, table0.GetCol(5).IsStatsInitialized())
```
Comment (winoros) [inline pkg/statistics/handle/handletest/statstest/stats_test.go:380]: column 1/2/3 is analyzed. col4 is not analyzed and has no record in storage.
col5 is created by `add column default value` so it has stats though it's not analyzed.

File: pkg/statistics/handle/handletest/initstats/load_stats_test.go [hunk 4]
```diff
@@ -88,7 +88,7 @@ func testConcurrentlyInitStats(t *testing.T) {
 		tk.MustQuery(fmt.Sprintf("explain select * from t%v where b = 1", i)).CheckNotContain("pseudo")
 	}
 	for i := 1; i < 10; i++ {
-		tk.MustQuery(fmt.Sprintf("explain select * from t%v where c = 1", i)).CheckNotContain("pseudo")
```
Comment (winoros) [inline pkg/statistics/handle/handletest/initstats/load_stats_test.go:91]: `pk = 1` will result in PointGet. Will not trigger the stats load.

**Review by hawkingrei — APPROVED** [review]

File: pkg/statistics/handle/bootstrap.go [hunk 5]
```diff
@@ -619,6 +638,15 @@ func (*Handle) initStatsBuckets4Chunk(cache statstypes.StatsCache, iter *chunk.I
 	}
 }
 
+func initStatsBucketsSQLGen(isPaging bool) string {
+	selectPrefix := "select /*+ ORDER_INDEX(mysql.stats_buckets,tbl) */ HIGH_PRIORITY table_id, is_index, hist_id, count, repeats, lower_bound, upper_bound, ndv from mysql.stats_buckets where is_index=1"
```
Comment (Rustin170506) [inline pkg/statistics/handle/bootstrap.go:642]: This seems different from the old one. Could you please explain why we need `is_index=1` here?

File: pkg/statistics/handle/bootstrap.go [hunk 6]
```diff
@@ -401,8 +411,17 @@ func (*Handle) initStatsTopN4Chunk(cache statstypes.StatsCache, iter *chunk.Iter
...
+	orderSuffix := " order by table_id"
+	if !isPaging {
+		return selectPrefix + orderSuffix
+	}
+	return selectPrefix + " and table_id >= %? and table_id < %?" + orderSuffix
+}
+
 func (h *Handle) initStatsTopN(cache statstypes.StatsCache, totalMemory uint64) error {
-	sql := "select /*+ ORDER_INDEX(mysql.stats_top_n,tbl)*/  HIGH_PRIORITY table_id, hist_id, value, count from mysql.stats_top_n where is_index = 1 order by table_id"
+	sql := initStatsTopNSQLGen(false)
```
Comment (Rustin170506) [inline pkg/statistics/handle/bootstrap.go:424]: This seems different from the old one. Could you please explain why we need to remove `is_index=1` here?

File: pkg/statistics/handle/bootstrap.go [hunk 7]
```diff
@@ -285,7 +298,7 @@ func (h *Handle) initStatsHistogramsLite(ctx context.Context, cache statstypes.S
 }
 
 func (h *Handle) initStatsHistograms(is infoschema.InfoSchema, cache statstypes.StatsCache) error {
-	sql := "select  /*+ ORDER_INDEX(mysql.stats_histograms,tbl)*/ HIGH_PRIORITY table_id, is_index, hist_id, distinct_count, version, null_count, cm_sketch, tot_col_size, stats_ver, correlation, flag, last_analyze_pos from mysql.stats_histograms order by table_id"
+	sql := initStatsHistogramsSQLGen(false)
```
Comment (Rustin170506) [inline pkg/statistics/handle/bootstrap.go:301]: This seems different from the old one. Could you please explain why we need to remove is_index=1 here?

**Review by Rustin170506 — COMMENTED** [review]
Could you please summarize the changes made in your PR description? It's a bit difficult to understand what is happening in the PR.

File: pkg/statistics/table.go [hunk 8]
```diff
@@ -819,16 +829,18 @@ func (t *Table) ColumnIsLoadNeeded(id int64, fullLoad bool) (*Column, bool, bool
 	if t.Pseudo {
 		return nil, false, false
 	}
-	// when we use non-lite init stats, it cannot init the stats for common columns.
-	// so we need to force to load the stats.
 	col, ok := t.columns[id]
+	hasAnalyzed := t.ColAndIdxExistenceMap.HasAnalyzed(id, false)
```
Comment (Rustin170506) [inline pkg/statistics/table.go:833]:
```suggestion
	hasAnalyzed := t.ColAndIdxExistenceMap.HasAnalyzed(id, false)
	col, ok := t.columns[id]
```

Placing the `ok` closer to the `ok` check would make the code easier to read.

File: pkg/statistics/handle/bootstrap.go (hunk 5 above)
Comment (winoros) [reply pkg/statistics/handle/bootstrap.go:642]: It's aligned with the non-concurrent version. This is to make the concurrent one and the non-concurrent one the same.

File: pkg/statistics/handle/bootstrap.go (hunk 6 above)
Comment (winoros) [reply pkg/statistics/handle/bootstrap.go:424]: the `is_index=1` is not removed

File: pkg/statistics/handle/bootstrap.go (hunk 7 above)
Comment (winoros) [reply pkg/statistics/handle/bootstrap.go:303]: If you search the caller of the `initStatsHistogramSQLGen` you will see that now `initStatsHistogramsLite`, `initStatsHistograms`, and `initStatsHistogramsConcurrency` use the same SQL gen.
Actually, the SQL is not changed for `initStatsHistogramsLite` and `initStatsHistograms`. It just aligned the `initStatsHistogramsConcurrency` with the other two.

File: pkg/statistics/handle/bootstrap.go [hunk 9]
```diff
@@ -244,15 +244,19 @@ func (h *Handle) initStatsHistograms4Chunk(is infoschema.InfoSchema, cache stats
...
 			lastAnalyzePos.Copy(&col.LastAnalyzePos)
 			table.SetCol(hist.ID, col)
 			table.ColAndIdxExistenceMap.InsertCol(colInfo.ID, statsVer != statistics.Version0 || ndv > 0 || nullCount > 0)
 			if statsVer != statistics.Version0 {
 				// The LastAnalyzeVersion is added by ALTER table so its value might be 0.
 				table.LastAnalyzeVersion = max(table.LastAnalyzeVersion, version)
+				// We will also set int primary key's loaded status to evicted.
+				col.StatsLoadedStatus = statistics.NewStatsAllEvictedStatus()
+			} else if col.NDV > 0 || col.NullCount > 0 {
+				// If NDV > 0 or NullCount > 0, we should set the stats loaded status to loaded. See comments in Column.StatsAvailable.
```
Comment (Rustin170506) [inline pkg/statistics/handle/bootstrap.go:256]: Do you mean by `set the stats loaded status to evicted`?

File: pkg/statistics/handle/bootstrap.go [hunk 10]
```diff
@@ -719,8 +755,10 @@ func (h *Handle) initStatsBucketsConcurrency(cache statstypes.StatsCache, totalM
 
 // InitStatsLite initiates the stats cache. The function is liter and faster than InitStats.
 // 1. Basic stats meta data is loaded.(count, modify count, etc.)
-// 2. Column/index stats are loaded. (only histogram)
+// 2. Column/index stats are marked whehter exists or not by initializing the table.ColAndIdxExistenceMap. (by reading mysql.stats_histograms)
```
Comment (Rustin170506) [inline pkg/statistics/handle/bootstrap.go:758]:
```suggestion
// 2. Column/index stats are marked as existing or not by initializing the table.ColAndIdxExistenceMap, based on data from mysql.stats_histograms
```

File: pkg/statistics/handle/bootstrap.go [hunk 11]
```diff
@@ -750,7 +788,10 @@ func (h *Handle) InitStatsLite(ctx context.Context) (err error) {
 
 // InitStats initiates the stats cache.
 // 1. Basic stats meta data is loaded.(count, modify count, etc.)
-// 2. Column/index stats are loaded. (histogram, topn, buckets, FMSketch)
+// 2. Index stats are fully loaded. (histogram, topn, buckets)
+// 3. Column stats are just marked whehter exists or not by initializing the table.ColAndIdxExistenceMap. (by reading mysql.stats_histograms)
```
Comment (Rustin170506) [inline pkg/statistics/handle/bootstrap.go:792]: ditto

File: pkg/statistics/handle/syncload/stats_syncload_test.go [hunk 12]
```diff
@@ -409,3 +409,53 @@ func TestSendLoadRequestsWaitTooLong(t *testing.T) {
...
+	// Try sync load.
+	tk.MustExec("select * from t where a >= 1 and b = 2 and c = 3 and d = 4")
+	statsTbl, ok = h.Get(tblInfo.ID)
+	require.True(t, ok)
+	require.True(t, statsTbl.GetCol(tblInfo.Columns[0].ID).IsFullLoad())
+	require.True(t, statsTbl.GetCol(tblInfo.Columns[1].ID).IsFullLoad())
+	require.True(t, statsTbl.GetCol(tblInfo.Columns[3].ID).IsFullLoad())
+	require.Nil(t, statsTbl.GetCol(tblInfo.Columns[2].ID))
+	_, loadNeeded, analyzed := statsTbl.ColumnIsLoadNeeded(tblInfo.Columns[2].ID, false)
+	// After the sync load. The column without any thing in storage should not be marked as loadNeeded any more.
```
Comment (Rustin170506) [inline pkg/statistics/handle/syncload/stats_syncload_test.go:458]: If we analyze column `c` next time, I assume we can load it correctly? Could you please also add a test step to include it?

**Review by Rustin170506 — COMMENTED** [review]
rest LGTM
Thank you!

File: pkg/statistics/handle/bootstrap.go (hunk 7 above)
Comment (Rustin170506) [reply pkg/statistics/handle/bootstrap.go:303]: Got it!

File: pkg/statistics/handle/bootstrap.go (hunk 5 above)
Comment (Rustin170506) [reply pkg/statistics/handle/bootstrap.go:650]: Got it.

File: pkg/statistics/handle/syncload/stats_syncload_test.go (hunk 12 above)
Comment (winoros) [reply pkg/statistics/handle/syncload/stats_syncload_test.go:458]: Yes, if we analyze it. After the sync load. It should return `loadNeeded=false, analyzed=true`

**Review by Rustin170506 — APPROVED** [review]
Thanks! :shipit:

**Pull request merged**